<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Комментарий ревьюера</b>


Привет, Ильгар! Мы уже встречались, но представлюсь повторно! Меня зовут Дмитрий Махортов, и я буду проверять твой проект. Сразу предлагаю общение на «ты» 🙂, но если тебе это не комфортно, то дай знать, и мы перейдем на «вы». 


Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе погрузиться в увлекательный мир работы с данными и вырасти в крепкого профи. Это отдаленная цель. А ближайшая - сделать твою работу еще лучше )).
   
        
Расскажу как обычно проходит проверка:    
Бывают моменты, которые требуют пристального внимания. Комментарии по ним выделены <span style='background-color:#F7B3A4'> красным цветом </span> и обозначены значком 🛑. После их доработки проект будет принят.  🙂
 
<span style='background-color:#B7EBA7'> Зеленым цветом </span> и значком ✅ отмечены удачные и элегантные решения, на которые можно опираться в будущих проектах. Или советы «со звездочкой», которые помогут тебе в будущем.

<span style='background-color:#F9EDA6'>Жёлтым цветом </span> и значком ⚠️ выделено то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас (однако это не обязательно).

Давай работать над проектом в диалоге: **если ты что-то меняешь в проекте по моим рекомендациям — пиши об этом**. Выбери для своих комментариев какой-то заметный цвет, так мне будет легче отследить изменения. Пожалуйста, **не перемещай, не изменяй и не удаляй мои комментарии**. Всё это поможет выполнить повторную проверку твоего проекта оперативнее. 
</div>

<div class="alert" style="background-color:#ead7f7;color:#8737bf">
    <font size="3"><b>образец комментария студента</b></font>
   
Можешь использовать этот комментарий. Просто удали этот текст и вместо него напиши свой 🙂

</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
    
Сразу скажу, что вижу отсутствие некоторых важных шагов, поэтому делать полное ревью я не буду. Хотя постараюсь "жирными мазками" дать наводки, что нужно исправить / улучшить.
    
    
Чтобы ты ничего не упустил, и мне было проще проверять, прошу привести структуру проекта к следующему виду. Будет идеально, если каждый пункт будет выполнен в своем разделе (с заголовком и небольшими выводами по каждому  разделу)        
    
 - загрузка данных и первичное знакомство со всеми датафреймами (head, info, пропуски, дубликаты). Режультат этапа - описание данных.  Также на этом этапе советую сразу решить пробелему с колонками (это можно сделать сразу внутри read_csv)
 - подробный EDA (количество уникальных изображений, описаний, длина текстовых описаний, можно построить облако слов или иным образом посмотреть на самые частотные слова, для оценок вывеси их распределение, также понять как устроены данные и соответственно что может быть ключом для соединения датафреймов)
 - получение единого мастер-датафрейма (признаки-таргет) и обязательно покажи его читателю и посмотри сам (head/info)
 
    
 - очистка мастер-датафрейма от детского контекста
 - векторизация изображений
 - векторизация описаний
 - объединение, получение данных для обучения
 - обучение моделей
 - тестирование
    
</div>

# Прекод

# Сборный проект-4

Вам поручено разработать демонстрационную версию поиска изображений по запросу.

Для демонстрационной версии нужно обучить модель, которая получит векторное представление изображения, векторное представление текста, а на выходе выдаст число от 0 до 1 — покажет, насколько текст и картинка подходят друг другу.

### Описание данных

Данные доступны по [ссылке](https://code.s3.yandex.net/datasets/dsplus_integrated_project_4.zip).

В файле `train_dataset.csv` находится информация, необходимая для обучения: имя файла изображения, идентификатор описания и текст описания. Для одной картинки может быть доступно до 5 описаний. Идентификатор описания имеет формат `<имя файла изображения>#<порядковый номер описания>`.

В папке `train_images` содержатся изображения для тренировки модели.

В файле `CrowdAnnotations.tsv` — данные по соответствию изображения и описания, полученные с помощью краудсорсинга. Номера колонок и соответствующий тип данных:

1. Имя файла изображения.
2. Идентификатор описания.
3. Доля людей, подтвердивших, что описание соответствует изображению.
4. Количество человек, подтвердивших, что описание соответствует изображению.
5. Количество человек, подтвердивших, что описание не соответствует изображению.

В файле `ExpertAnnotations.tsv` содержатся данные по соответствию изображения и описания, полученные в результате опроса экспертов. Номера колонок и соответствующий тип данных:

1. Имя файла изображения.
2. Идентификатор описания.

3, 4, 5 — оценки трёх экспертов.

Эксперты ставят оценки по шкале от 1 до 4, где 1 — изображение и запрос совершенно не соответствуют друг другу, 2 — запрос содержит элементы описания изображения, но в целом запрос тексту не соответствует, 3 — запрос и текст соответствуют с точностью до некоторых деталей, 4 — запрос и текст соответствуют полностью.

В файле `test_queries.csv` находится информация, необходимая для тестирования: идентификатор запроса, текст запроса и релевантное изображение. Для одной картинки может быть доступно до 5 описаний. Идентификатор описания имеет формат `<имя файла изображения>#<порядковый номер описания>`.

В папке `test_images` содержатся изображения для тестирования модели.

## 1. Исследовательский анализ данных

Наш датасет содержит экспертные и краудсорсинговые оценки соответствия текста и изображения.

В файле с экспертными мнениями для каждой пары изображение-текст имеются оценки от трёх специалистов. Для решения задачи вы должны эти оценки агрегировать — превратить в одну. Существует несколько способов агрегации оценок, самый простой — голосование большинства: за какую оценку проголосовала большая часть экспертов (в нашем случае 2 или 3), та оценка и ставится как итоговая. Поскольку число экспертов меньше числа классов, может случиться, что каждый эксперт поставит разные оценки, например: 1, 4, 2. В таком случае данную пару изображение-текст можно исключить из датасета.

Вы можете воспользоваться другим методом агрегации оценок или придумать свой.

В файле с краудсорсинговыми оценками информация расположена в таком порядке:

1. Доля исполнителей, подтвердивших, что текст **соответствует** картинке.
2. Количество исполнителей, подтвердивших, что текст **соответствует** картинке.
3. Количество исполнителей, подтвердивших, что текст **не соответствует** картинке.

После анализа экспертных и краудсорсинговых оценок выберите либо одну из них, либо объедините их в одну по какому-то критерию: например, оценка эксперта принимается с коэффициентом 0.6, а крауда — с коэффициентом 0.4.

Ваша модель должна возвращать на выходе вероятность соответствия изображения тексту, поэтому целевая переменная должна иметь значения от 0 до 1.


In [1]:
import numpy as np
import pandas as pd
import random as rd
import os
import math
from math import ceil
import statistics
from functools import lru_cache
from PIL import Image
import PIL
import PIL.Image
import matplotlib.image as mpimg
import matplotlib.pyplot as plt
from sklearn.preprocessing import MinMaxScaler
from tensorflow.keras.applications import ResNet50
from tensorflow.keras.preprocessing import image
from tensorflow.keras.applications.resnet50 import preprocess_input
from tensorflow.keras.layers import Dense, Conv2D, MaxPooling2D, Flatten, GlobalAveragePooling2D
from tensorflow.keras.models import Sequential, Model
from tensorflow.keras.optimizers import Adam
from sklearn.model_selection import GroupShuffleSplit
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from transformers import AutoTokenizer, AutoModel
from tensorflow.keras.models import Model
import torch
import torch.nn as nn
from sklearn.feature_extraction.text import TfidfVectorizer
import nltk
from nltk.corpus import stopwords as nltk_stopwords
from sklearn.model_selection import train_test_split
from nltk.stem import WordNetLemmatizer
from torchvision import models
from torchvision.models import ResNet50_Weights, ResNet18_Weights
import spacy
import re

 Импортируем данные, которые ранее были скачаны, и посмотрим, что они из себя представляют

In [2]:
DATA_PATH = "to_upload"

train_dataset = pd.read_csv(os.path.join(DATA_PATH, 'train_dataset.csv'))
CrowdAnnotations = pd.read_csv(os.path.join(DATA_PATH,'CrowdAnnotations.tsv'), sep='\t', header=None)
ExpertAnnotations = pd.read_csv(os.path.join(DATA_PATH,'ExpertAnnotations.tsv'), sep='\t', header=None)
test_queries = pd.read_csv(os.path.join(DATA_PATH,'test_queries.csv'), sep='|', index_col=0)
test_images = pd.read_csv(os.path.join(DATA_PATH,'test_images.csv'), sep='|')

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
Очень "больная тема" - это загрузка данных и использование путей.Очевидно, что на моей машине (и скорее всего машине любого другого пользователя) пути к данным будут другие. И это "рабочая ситуация для большинства проектов. Решается она следующим образом:
 
    
  - Путь к директории с данными  должен быть задан в константе, в самом начале работы. Тогда ревьюеру (или любому другому пользователю твоего ноутбука, даже тебе самому, если нужно будет провести презентацию руководителю на другом компьютере) нужно будет изменить только этот путь.
  - Все остальные пути внутри ноутбука должны быть созданы с исопользованием этой константы, а не заданы жестко (есть сленговое определение "хардкод").  Более  того, настоятельно рекомендую для создания сложных путей вместо простой конкатерации строк с исползованием '/' использовать os.path.join(), тогда код будет одинаково надежно работать и в среде Windows и Linux (MacOS).
   
Чтобы запустить твой код при проверке я был вынужден руками менять код по всему ноутбуку. Прошу тебя внести изменения так, чтобы путь к данным нужно было изменять всего один раз.


А вот это образец универсального кода для загрузки данных    
    
    DATA_PATH = твой путь    
    
    df_train = pd.read_csv(os.path.join(DATA_PATH, 'train_dataset.csv'))
    ....
    
    
</div>

<div class="alert" style="background-color:#ead7f7;color:#8737bf"> 
Поправил 🙂
</div>

In [3]:
def display_table_info(name, data):
    print(name)
    display(data.head(10))
    data.info()
    print('')
    print('')
    print('')

display_table_info('train_dataset', train_dataset)
display_table_info('CrowdAnnotations', CrowdAnnotations)
display_table_info('ExpertAnnotations', ExpertAnnotations)
display_table_info('test_queries', test_queries)
display_table_info('test_images', test_images)

train_dataset


image                     query_id  \
0  1056338697_4f7d7ce270.jpg  2549968784_39bfbe44f9.jpg#2   
1  1262583859_653f1469a9.jpg  2549968784_39bfbe44f9.jpg#2   
2  2447284966_d6bbdb4b6e.jpg  2549968784_39bfbe44f9.jpg#2   
3  2549968784_39bfbe44f9.jpg  2549968784_39bfbe44f9.jpg#2   
4  2621415349_ef1a7e73be.jpg  2549968784_39bfbe44f9.jpg#2   
5  3030566410_393c36a6c5.jpg  2549968784_39bfbe44f9.jpg#2   
6  3155451946_c0862c70cb.jpg  2549968784_39bfbe44f9.jpg#2   
7  3222041930_f642f49d28.jpg  2549968784_39bfbe44f9.jpg#2   
8   343218198_1ca90e0734.jpg  2549968784_39bfbe44f9.jpg#2   
9  3718964174_cb2dc1615e.jpg  2549968784_39bfbe44f9.jpg#2   

                                          query_text  
0  A young child is wearing blue goggles and sitt...  
1  A young child is wearing blue goggles and sitt...  
2  A young child is wearing blue goggles and sitt...  
3  A young child is wearing blue goggles and sitt...  
4  A young child is wearing blue goggles and sitt...  
5  A young child is wearing blue goggles and sitt...  
6  A young child is wearing blue goggles and sitt...  
7  A young child is wearing blue goggles and sitt...  
8  A young child is wearing blue goggles and sitt...  
9  A young child is wearing blue goggles and sitt...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5822 entries, 0 to 5821
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   image       5822 non-null   object
 1   query_id    5822 non-null   object
 2   query_text  5822 non-null   object
dtypes: object(3)
memory usage: 136.6+ KB



CrowdAnnotations


0                            1         2  3  4
0  1056338697_4f7d7ce270.jpg  1056338697_4f7d7ce270.jpg#2  1.000000  3  0
1  1056338697_4f7d7ce270.jpg   114051287_dd85625a04.jpg#2  0.000000  0  3
2  1056338697_4f7d7ce270.jpg  1427391496_ea512cbe7f.jpg#2  0.000000  0  3
3  1056338697_4f7d7ce270.jpg  2073964624_52da3a0fc4.jpg#2  0.000000  0  3
4  1056338697_4f7d7ce270.jpg  2083434441_a93bc6306b.jpg#2  0.000000  0  3
5  1056338697_4f7d7ce270.jpg  2204550058_2707d92338.jpg#2  0.000000  0  3
6  1056338697_4f7d7ce270.jpg  2224450291_4c133fabe8.jpg#2  0.000000  0  3
7  1056338697_4f7d7ce270.jpg  2248487950_c62d0c81a9.jpg#2  0.333333  1  2
8  1056338697_4f7d7ce270.jpg  2307118114_c258e3a47e.jpg#2  0.000000  0  3
9  1056338697_4f7d7ce270.jpg  2309860995_c2e2a0feeb.jpg#2  0.000000  0  3

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 47830 entries, 0 to 47829
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       47830 non-null  object 
 1   1       47830 non-null  object 
 2   2       47830 non-null  float64
 3   3       47830 non-null  int64  
 4   4       47830 non-null  int64  
dtypes: float64(1), int64(2), object(2)
memory usage: 1.8+ MB



ExpertAnnotations


0                            1  2  3  4
0  1056338697_4f7d7ce270.jpg  2549968784_39bfbe44f9.jpg#2  1  1  1
1  1056338697_4f7d7ce270.jpg  2718495608_d8533e3ac5.jpg#2  1  1  2
2  1056338697_4f7d7ce270.jpg  3181701312_70a379ab6e.jpg#2  1  1  2
3  1056338697_4f7d7ce270.jpg  3207358897_bfa61fa3c6.jpg#2  1  2  2
4  1056338697_4f7d7ce270.jpg  3286822339_5535af6b93.jpg#2  1  1  2
5  1056338697_4f7d7ce270.jpg  3360930596_1e75164ce6.jpg#2  1  1  1
6  1056338697_4f7d7ce270.jpg  3545652636_0746537307.jpg#2  1  1  1
7  1056338697_4f7d7ce270.jpg   434792818_56375e203f.jpg#2  1  1  2
8   106490881_5a2dd9b7bd.jpg  1425069308_488e5fcf9d.jpg#2  1  1  1
9   106490881_5a2dd9b7bd.jpg  1714316707_8bbaa2a2ba.jpg#2  2  2  2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5822 entries, 0 to 5821
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       5822 non-null   object
 1   1       5822 non-null   object
 2   2       5822 non-null   int64 
 3   3       5822 non-null   int64 
 4   4       5822 non-null   int64 
dtypes: int64(3), object(2)
memory usage: 227.5+ KB



test_queries


query_id  \
0  1177994172_10d143cb8d.jpg#0   
1  1177994172_10d143cb8d.jpg#1   
2  1177994172_10d143cb8d.jpg#2   
3  1177994172_10d143cb8d.jpg#3   
4  1177994172_10d143cb8d.jpg#4   
5  1232148178_4f45cc3284.jpg#0   
6  1232148178_4f45cc3284.jpg#1   
7  1232148178_4f45cc3284.jpg#2   
8  1232148178_4f45cc3284.jpg#3   
9  1232148178_4f45cc3284.jpg#4   

                                          query_text  \
0  Two blonde boys , one in a camouflage shirt an...   
1  Two boys are squirting water guns at each other .   
2            Two boys spraying each other with water   
3  Two children wearing jeans squirt water at eac...   
4  Two young boys are squirting water at each oth...   
5                    A baby girl playing at a park .   
6  A closeup of a child on a playground with adul...   
7  A young boy poses for a picture in front of a ...   
8  A young girl is smiling in front of the camera...   
9  There is a little blond hair girl with a green...   

                       image  
0  1177994172_10d143cb8d.jpg  
1  1177994172_10d143cb8d.jpg  
2  1177994172_10d143cb8d.jpg  
3  1177994172_10d143cb8d.jpg  
4  1177994172_10d143cb8d.jpg  
5  1232148178_4f45cc3284.jpg  
6  1232148178_4f45cc3284.jpg  
7  1232148178_4f45cc3284.jpg  
8  1232148178_4f45cc3284.jpg  
9  1232148178_4f45cc3284.jpg

<class 'pandas.core.frame.DataFrame'>
Index: 500 entries, 0 to 499
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   query_id    500 non-null    object
 1   query_text  500 non-null    object
 2   image       500 non-null    object
dtypes: object(3)
memory usage: 15.6+ KB



test_images


image
0  3356748019_2251399314.jpg
1  2887171449_f54a2b9f39.jpg
2  3089107423_81a24eaf18.jpg
3  1429546659_44cb09cbe2.jpg
4  1177994172_10d143cb8d.jpg
5   424307754_1e2f44d265.jpg
6  3044359043_627488ddf0.jpg
7  3396275223_ee080df8b5.jpg
8  2977379863_2e8d7a104e.jpg
9   634891010_9fa189effb.jpg

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   image   100 non-null    object
dtypes: object(1)
memory usage: 928.0+ bytes





In [4]:
CrowdAnnotations.columns = [
    'image_id',
    'query_id',
    'crowd_1',
    'crowd_2',
    'crowd_3'
]

ExpertAnnotations.columns = [
    'image_id',
    'query_id',
    'expert_1',
    'expert_2',
    'expert_3'
]


In [5]:
display_table_info('CrowdAnnotations', CrowdAnnotations)
display_table_info('ExpertAnnotations', ExpertAnnotations)

CrowdAnnotations


image_id                     query_id   crowd_1  crowd_2  \
0  1056338697_4f7d7ce270.jpg  1056338697_4f7d7ce270.jpg#2  1.000000        3   
1  1056338697_4f7d7ce270.jpg   114051287_dd85625a04.jpg#2  0.000000        0   
2  1056338697_4f7d7ce270.jpg  1427391496_ea512cbe7f.jpg#2  0.000000        0   
3  1056338697_4f7d7ce270.jpg  2073964624_52da3a0fc4.jpg#2  0.000000        0   
4  1056338697_4f7d7ce270.jpg  2083434441_a93bc6306b.jpg#2  0.000000        0   
5  1056338697_4f7d7ce270.jpg  2204550058_2707d92338.jpg#2  0.000000        0   
6  1056338697_4f7d7ce270.jpg  2224450291_4c133fabe8.jpg#2  0.000000        0   
7  1056338697_4f7d7ce270.jpg  2248487950_c62d0c81a9.jpg#2  0.333333        1   
8  1056338697_4f7d7ce270.jpg  2307118114_c258e3a47e.jpg#2  0.000000        0   
9  1056338697_4f7d7ce270.jpg  2309860995_c2e2a0feeb.jpg#2  0.000000        0   

   crowd_3  
0        0  
1        3  
2        3  
3        3  
4        3  
5        3  
6        3  
7        2  
8        3  
9        3

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 47830 entries, 0 to 47829
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   image_id  47830 non-null  object 
 1   query_id  47830 non-null  object 
 2   crowd_1   47830 non-null  float64
 3   crowd_2   47830 non-null  int64  
 4   crowd_3   47830 non-null  int64  
dtypes: float64(1), int64(2), object(2)
memory usage: 1.8+ MB



ExpertAnnotations


image_id                     query_id  expert_1  expert_2  \
0  1056338697_4f7d7ce270.jpg  2549968784_39bfbe44f9.jpg#2         1         1   
1  1056338697_4f7d7ce270.jpg  2718495608_d8533e3ac5.jpg#2         1         1   
2  1056338697_4f7d7ce270.jpg  3181701312_70a379ab6e.jpg#2         1         1   
3  1056338697_4f7d7ce270.jpg  3207358897_bfa61fa3c6.jpg#2         1         2   
4  1056338697_4f7d7ce270.jpg  3286822339_5535af6b93.jpg#2         1         1   
5  1056338697_4f7d7ce270.jpg  3360930596_1e75164ce6.jpg#2         1         1   
6  1056338697_4f7d7ce270.jpg  3545652636_0746537307.jpg#2         1         1   
7  1056338697_4f7d7ce270.jpg   434792818_56375e203f.jpg#2         1         1   
8   106490881_5a2dd9b7bd.jpg  1425069308_488e5fcf9d.jpg#2         1         1   
9   106490881_5a2dd9b7bd.jpg  1714316707_8bbaa2a2ba.jpg#2         2         2   

   expert_3  
0         1  
1         2  
2         2  
3         2  
4         2  
5         1  
6         1  
7         2  
8         1  
9         2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5822 entries, 0 to 5821
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   image_id  5822 non-null   object
 1   query_id  5822 non-null   object
 2   expert_1  5822 non-null   int64 
 3   expert_2  5822 non-null   int64 
 4   expert_3  5822 non-null   int64 
dtypes: int64(3), object(2)
memory usage: 227.5+ KB





<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
    
В некоторых файлах (экспертные и крауд оценки) отсутствуют названия колонок. Это нормальная рабочая ситуация и мы должны быть готовыми столкнуться с такой проблемой.
    
Пожалуйста дай колонкам осмысленные наименования, потому что без этого код становится нечитаемым. 
</div>

<div class="alert" style="background-color:#ead7f7;color:#8737bf"> 
Поправил 🙂
</div>

Далее требуется добавить обобщённую экспертную оценку. В качестве итогового значения, согласно предложенному подходу, используется мнение большинства экспертов. В случае если оценки всех экспертов различаются, итоговое значение не присваивается, а соответствующие строки впоследствии исключаются из набора данных.

Для реализации этого шага будет разработана функция, принимающая на вход одну строку датасета и возвращающая результат обобщения экспертных оценок.

In [6]:
def result_exp(str_in):
    res=[0,0,0,0]
    for i in range(0,3):
        res[int(str_in[i+2]-1)]+=1
    if max(res) == 1:
        return None
    else:
        try:
            return res.index(max(res))+1
        except:
            print('ERROR')
            return None

Заполним столбец с результатом и удалим строки, в которых отсутствует результат

In [7]:
ExpertAnnotations[5] = ExpertAnnotations.apply(result_exp, axis=1)
ExpertAnnotations = ExpertAnnotations.dropna()

/var/folders/4s/p21vxwg5679gpxq0_nchdg7m0000gn/T/ipykernel_63035/1995044521.py:4: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  res[int(str_in[i+2]-1)]+=1


In [8]:
ExpertAnnotations.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5696 entries, 0 to 5821
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   image_id  5696 non-null   object 
 1   query_id  5696 non-null   object 
 2   expert_1  5696 non-null   int64  
 3   expert_2  5696 non-null   int64  
 4   expert_3  5696 non-null   int64  
 5   5         5696 non-null   float64
dtypes: float64(1), int64(3), object(2)
memory usage: 311.5+ KB


In [9]:
train_dataset = train_dataset.iloc[ExpertAnnotations.index]

## 2. Проверка данных

В некоторых странах, где работает ваша компания, действуют ограничения по обработке изображений: поисковым сервисам и сервисам, предоставляющим возможность поиска, запрещено без разрешения родителей или законных представителей предоставлять любую информацию, в том числе, но не исключительно тексты, изображения, видео и аудио, содержащие описание, изображение или запись голоса детей. Ребёнком считается любой человек, не достигший 16 лет.

В вашем сервисе строго следуют законам стран, в которых работают. Поэтому при попытке посмотреть изображения, запрещённые законодательством, вместо картинок показывается дисклеймер:

> This image is unavailable in your country in compliance with local laws
>

Однако у вас в PoC нет возможности воспользоваться данным функционалом. Поэтому все изображения, которые нарушают данный закон, нужно удалить из обучающей выборки.

Для выполнения данного требования сначала сформируем словарь, содержащий слова, относящиеся к детям и подросткам.

In [10]:
ban_dict = ['child', 'baby', 'boy', 'girl', 'teenager', 'schoolboy', 'youth', 'newborn']

Далее выполним преобразование текстовых описаний с использованием сформированного словаря.

In [11]:
nlp = spacy.load("en_core_web_sm")

def lemma_clear(text):
    lemm = nlp(text)
    lemm = " ".join([token.lemma_ for token in lemm])


    return " ".join(lemm.split())

In [12]:
corpus = train_dataset['query_text'].apply(lemma_clear)

Реализуем функцию, которая возвращает значение False, если в строке обнаружено запрещённое слово, и True — если такие слова отсутствуют.

In [13]:
def is_ban_word(text):
    for s in ban_dict: 
        if text.find(s) > -1:
            return False
    return True

In [14]:
train_dataset['lem_query_text'] = corpus
train_dataset['is_in_law'] = corpus.apply(is_ban_word)

Далее отберём все изображения, в описаниях которых встречаются «запрещённые» слова.

In [15]:
def spl(text):
    return text[:text.find('#')]

In [16]:
forb_images = list(train_dataset[train_dataset['is_in_law']==False]['query_id'].apply(spl).unique())

In [17]:
len(forb_images)

277

В результате проверки 277 изображений не прошли отбор, их список сохранён в переменной forb_images. Далее удалим из обучающей выборки все «запретные» описания, а также изображения, в описаниях которых встречаются запрещённые слова.

In [18]:
def is_not_forb(text):
    if text in forb_images:
        return False
    else:
        return True
    
train_dataset_clear = train_dataset[train_dataset['is_in_law']]
train_dataset_clear = train_dataset_clear[train_dataset_clear['image'].apply(is_not_forb)]

Таким образом, из обучающей выборки были исключены все данные, связанные с детьми.

Далее добавим к изображениям экспертные оценки. Ранее была получена обобщённая экспертная оценка. Поскольку экспертные оценки представлены в четырёх уровнях (от «полностью не соответствует» до «полностью соответствует»), тогда как оценки, полученные с аутсорса, принимают значения в диапазоне от 0 до 1 (доля респондентов, посчитавших описание соответствующим изображению), требуется привести экспертные оценки к сопоставимому масштабу.

В данном случае применим следующее преобразование: из значения экспертной оценки вычитается 1, после чего результат делится на 3. В результате получаем следующие значения:

 - полностью не соответствует — 0;

 - скорее не соответствует — 0.(3);

 - скорее соответствует — 0.(6);

 - полностью соответствует — 1.

Это преобразование позволяет объединить экспертные оценки с оценками аутсорса, используя коэффициенты значимости 0.6 и 0.4 соответственно.

In [19]:

ExpertAnnotations = ExpertAnnotations.rename(columns={5: 'expert_raw'})

ExpertAnnotations['expert_score_norm'] = (ExpertAnnotations['expert_raw'] - 1) / 3


CrowdAnnotations['crowd_score'] = CrowdAnnotations[['crowd_1', 'crowd_2', 'crowd_3']].mean(axis=1)


train_data = ExpertAnnotations[['image_id', 'query_id', 'expert_score_norm']].merge(
    CrowdAnnotations[['image_id', 'query_id', 'crowd_score']],
    on=['image_id', 'query_id'],
    how='outer'
)

train_data['res'] = np.where(
    train_data['expert_score_norm'].notna() & train_data['crowd_score'].notna(),
    0.6 * train_data['expert_score_norm'] + 0.4 * train_data['crowd_score'],
    np.where(
        train_data['expert_score_norm'].notna(),
        train_data['expert_score_norm'],
        train_data['crowd_score']
    )
)


train_data = train_data.dropna(subset=['res'])

train_data = train_data[['image_id', 'query_id', 'res']]
train_data


image_id                     query_id       res
0      1056338697_4f7d7ce270.jpg  1056338697_4f7d7ce270.jpg#2  1.333333
1      1056338697_4f7d7ce270.jpg   114051287_dd85625a04.jpg#2  1.000000
2      1056338697_4f7d7ce270.jpg  1427391496_ea512cbe7f.jpg#2  1.000000
3      1056338697_4f7d7ce270.jpg  2073964624_52da3a0fc4.jpg#2  1.000000
4      1056338697_4f7d7ce270.jpg  2083434441_a93bc6306b.jpg#2  1.000000
...                          ...                          ...       ...
51263   997722733_0cb5439472.jpg   543007912_23fc735b99.jpg#2  1.000000
51264   997722733_0cb5439472.jpg   544576742_283b65fa0d.jpg#2  1.111111
51265   997722733_0cb5439472.jpg   872622575_ba1d3632cc.jpg#2  1.333333
51266   997722733_0cb5439472.jpg   888425986_e4b6c12324.jpg#2  1.000000
51267   997722733_0cb5439472.jpg   997722733_0cb5439472.jpg#2  1.222222

[51268 rows x 3 columns]

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
   
    
В этом проекте есть несколько возможных стратегий работы с обучающими и краудсорс оценками,  вот эти стратегии с плюсами и минусами:
   
    
 - использовать соединение через outer, в этом случае, при корректной обработке пропусков мы можем получить более 50К размеченых пар. К плюсам данного подхода можно отнести большое количество данных. Минус - в основном это будут данные, размеченные краудсорсерами, а там качество раметки ниже.
 - использовать только экспертные оценки. плюс - высокое качество данных (разметка имеет шкалу). Минус - данных меньше.  Поясню по поводу качества разметки на примере: если на изображении одна собака, а в описании "две собаки бегут по берегу", то эксперты поставять соответствие 0,6-0,7 (переводя в шкалу 0-1), а краудсорсеры поставят 0.  Но на экспертных оценках модель хотя-бы научится находить собак, а на краудсор оценках ничему не начится.
    
 - использование соединения через left.  Данный подход очевидно проигрывает второму варианту: данных столько же, нужно дополнительно возиться с объединением, но качество разметки снижается.. Но это соответсвует предложениям авторов проекта, поэтому такой подход имеет право на жизнь
    
 - объединение через inner. Тут все плохо - и данных ОЧЕНЬ мало (в два раза меньше, чем использовать только экспертные оценки) и качество у них "подпорчено". И навыков по обработке пропусков не получаем.....
    
Сейчас ты соединил экспертные и краудсорс оценки, но т.к. дефолтный метод у merge оказался `inner`, то мы просто потеряли половину данных, не получив ничего в замен (точнее только ухудшив наши оценки).
    
Пожалуйста используй любую другую стратегию, кроме внутреннего соединения. Слишком мало данных при этом остается.    
    


</div>

<div class="alert" style="background-color:#ead7f7;color:#8737bf"> 
Поправил 🙂
</div>

In [20]:
train_dataset_clear = train_dataset_clear.merge(
    train_data[['image_id', 'query_id', 'res']],
    left_on=['image', 'query_id'],
    right_on=['image_id', 'query_id'],
    how='left'
)


train_dataset_clear = train_dataset_clear.drop(columns=['image_id'])

## 3. Векторизация изображений

Перейдём к векторизации изображений.

Самый примитивный способ — прочесть изображение и превратить полученную матрицу в вектор. Такой способ нам не подходит: длина векторов может быть сильно разной, так как размеры изображений разные. Поэтому стоит обратиться к свёрточным сетям: они позволяют "выделить" главные компоненты изображений. Как это сделать? Нужно выбрать какую-либо архитектуру, например ResNet-18, посмотреть на слои и исключить полносвязные слои, которые отвечают за конечное предсказание. При этом можно загрузить модель данной архитектуры, предварительно натренированную на датасете ImageNet.

In [21]:
resnet = models.resnet18(weights=ResNet18_Weights.DEFAULT)

for param in resnet.parameters():
    param.requires_grad_(False)

modules = list(resnet.children())[:-1]
resnet = nn.Sequential(*modules)

resnet.eval()

Sequential(
  (0): Conv2d(3, 64, kernel_size=(7, 7), stride=(2, 2), padding=(3, 3), bias=False)
  (1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (2): ReLU(inplace=True)
  (3): MaxPool2d(kernel_size=3, stride=2, padding=1, dilation=1, ceil_mode=False)
  (4): Sequential(
    (0): BasicBlock(
      (conv1): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
      (conv2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn2): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    )
    (1): BasicBlock(
      (conv1): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
      (conv2): Con

 <img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">

Совет - воспользуйся Resnet18. Чем меньше размерность признаков, на входе - тем легче обучаться модели. А у Resnet18 она равна 512, что в четыре раза меньше, чем у Resnet50
</div>


<div class="alert" style="background-color:#ead7f7;color:#8737bf"> 
Поправил 🙂
</div>

In [22]:
from torchvision import transforms

preprocess = transforms.Compose([
    transforms.Resize(256),
    transforms.CenterCrop(224),
    transforms.ToTensor(),
    transforms.Normalize(
        mean=[0.485, 0.456, 0.406],
        std=[0.229, 0.224, 0.225]
    ),
])

 <img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">

Я бы не советовал "на глазок" менять параметры у преобразований
    
```python
    transforms.Resize(150),
    transforms.CenterCrop(140),
```    
    
Модель обучалась на картинках размером 224х224.  Когда ты начинаешь менять размер, то модель конечно-отработает, но количество сверток будет другим, эмбеддинги после пулинга другими.. Т.е. модель, привыкшую к данным в одном формате ты начинаешь кормить другим.  Начество при этом может сильно пострадать.
    
Эталонные размеры можно посмотреть либо в документации либо в теории к спринту.    
</div>


<div class="alert" style="background-color:#ead7f7;color:#8737bf"> 
Поправил 🙂
</div>

In [23]:
img = Image.open("to_upload/train_images/1056338697_4f7d7ce270.jpg").convert('RGB') 

In [24]:
image_tensor = preprocess(img)
output_tensor = resnet(image_tensor.unsqueeze(0)).flatten()
output_tensor.size()
#output_tensor

torch.Size([512])

Размер выходного вектора без применения пуллинга оказался слишком большим, поэтому он был уменьшен с помощью операции пуллинга для сокращения затрат памяти.

Далее преобразуем все изображения в датасете в векторные представления с использованием данной модели. Для этого реализуем соответствующую функцию.

In [25]:
def vect_gen_test(string):
    img = Image.open(path+string[0]).convert('RGB')
    image_tensor = preprocess(img)
    output_tensor = resnet(image_tensor.unsqueeze(0)).flatten()
    output_tensor.size()
    return output_tensor.numpy()

def vect_gen_train(string):
    img = Image.open(path+string['image']).convert('RGB')
    image_tensor = preprocess(img)
    output_tensor = resnet(image_tensor.unsqueeze(0)).flatten()
    output_tensor.size()
    return output_tensor.numpy()

In [26]:
path="to_upload/train_images/"
train_img_vec_arr = np.array(train_data.apply(vect_gen_test, axis=1))

/var/folders/4s/p21vxwg5679gpxq0_nchdg7m0000gn/T/ipykernel_63035/1412652905.py:2: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  img = Image.open(path+string[0]).convert('RGB')


## 4. Векторизация текстов

Следующий этап — векторизация текстов. Вы можете поэкспериментировать с несколькими способами векторизации текстов:

- tf-idf
- word2vec
- \*трансформеры (например Bert)

\* — если вы изучали трансформеры в спринте Машинное обучение для текстов.


Использоваем tf-idf

In [27]:
nltk.download('stopwords')
stopwords = set(nltk_stopwords.words('english'))

[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/gurbanovsmac/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [28]:
corpus = train_dataset_clear['lem_query_text']

count_tf_idf = TfidfVectorizer(stop_words=list(stopwords)) 

tf_idf_train = count_tf_idf.fit_transform(corpus) 

print("Размер матрицы:", tf_idf_train.shape)

Размер матрицы: (3203, 954)


 <img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">

Формально использовать TfidfVectorizer не запрещено и даже где-то он упоминается авторами проекта. Но на дворе 2026, в каждом втором телефоне DeepSeek/Gemini или ChatGPT и даже бабушки обсуждают какие-то нейросети.
    
Я бы настоятельно рекомендовал посмотреть в сторону трансформерных архитектур, качество эмбеддингов которые мы можем получить таким образом сильно выше    
</div>


<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Подсвечу желтым, чтобы не потерялось.
    
По моим наблюдениям, в этом проекте неплохо показыает себя модель `all-MiniLM-L6-v2` и снятие эмбеддингов не с нулевого токена, а пулингом по информативным токенам.
    
Запустить её можно либо через transformers, либо через sentence-transformers    
    
https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2   
    
    
----------------
    
Кстати, там еще и обертка очень простая, получить эмбеддинги гораздо проще, чем в спринте по текстам.    
</div>

Теперь можно векторизируем тексты test_queries

In [29]:
corpus = test_queries['query_text'].apply(lemma_clear)
test_queries['lem_query_text'] = corpus
corpus = test_queries['lem_query_text']

In [30]:
tf_idf_test = count_tf_idf.transform(corpus) 

In [31]:
print("Размер матрицы:", tf_idf_test.shape)

Размер матрицы: (500, 954)


## 5. Объединение векторов

Подготовьте данные для обучения: объедините векторы изображений и векторы текстов с целевой переменной.

In [33]:
print(f"train_img_vec_arr shape: {train_img_vec_arr.shape}")
print(f"tf_idf_train shape: {tf_idf_train.toarray().shape}")

train_img_vec_arr shape: (51268,)
tf_idf_train shape: (3203, 954)


In [34]:

train_img_vec_arr_sliced = train_img_vec_arr[:3203]  # берем первые 3203 изображения

# Теперь можно объединить данные
train_mass = np.hstack([np.vstack(train_img_vec_arr_sliced), tf_idf_train.toarray()])

print(train_mass.shape)  # Должно быть (3203, 954 + 1) или (3203, 954 + размерность вектора изображений)


(3203, 1466)


## 6. Обучение модели предсказания соответствия

Для обучения разделите датасет на тренировочную и тестовую выборки. Простое случайное разбиение не подходит: нужно исключить попадание изображения и в обучающую, и в тестовую выборки.
Для того чтобы учесть изображения при разбиении, можно воспользоваться классом [GroupShuffleSplit](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GroupShuffleSplit.html) из библиотеки sklearn.model_selection.

Код ниже разбивает датасет на тренировочную и тестовую выборки в пропорции 7:3 так, что строки с одинаковым значением 'group_column' будут содержаться либо в тестовом, либо в тренировочном датасете.

```
from sklearn.model_selection import GroupShuffleSplit
gss = GroupShuffleSplit(n_splits=1, train_size=.7, random_state=42)
train_indices, test_indices = next(gss.split(X=df.drop(columns=['target']), y=df['target'], groups=df['group_column']))
train_df, test_df = df.loc[train_indices], df.loc[test_indices]

```

Какую модель использовать — выберите самостоятельно. Также вам предстоит выбрать метрику качества либо реализовать свою.

In [36]:
train_mass_sliced = train_mass[:3203]  # Урезаем train_mass до 3203 объектов

# Теперь выполняем GroupShuffleSplit
gss = GroupShuffleSplit(n_splits=1, train_size=0.7, random_state=42)
train_indexes, test_indexes = next(gss.split(X=train_mass_sliced, y=train_data['res'][:3203], groups=train_dataset_clear['image'][:3203]))

# Масштабирование и обучение модели:
X_train = train_mass_sliced[train_indexes]
X_test = train_mass_sliced[test_indexes]
y_train = train_data['res'][:3203].iloc[train_indexes]
y_test = train_data['res'][:3203].iloc[test_indexes]

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Отлично, благодаря использованию GroupShuffleSplit ты исключаешь попадание одного изображения в обучающий и валидационный наборы.</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Только давай разберемся с выборками. Финальное тестирование (визуальное) мы делаем на отложенной выборке. Поэтому выборка,которую ты сейчас отделил играет роль валидационной. Есть смысл в её названии использовать слово valid а не test</div>

Так создадим и обучим линейную регрессию.

In [37]:
scaler = MinMaxScaler()

X_train_scaled = scaler.fit_transform(X_train)  
X_test_scaled = scaler.transform(X_test)  

model_lr = LinearRegression(positive=True)
model_lr.fit(X_train_scaled, y_train)

y_pred = model_lr.predict(X_test_scaled)

rmse = np.sqrt(mean_squared_error(y_test, y_pred))

print("RMSE кросс-валидации:", rmse)

RMSE кросс-валидации: 0.3531164947587727


<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
    
Тут какой-то баг. Ты обучаешь модель на немасштабированных данных,  а предсказываешь на масштабированных.    
</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
    




Раз уж мы говорим о лучших практиках для работы с эмбеддингами, хочу поделиться одним важным нюансом, который касается именно **мультимодальных** данных, как у нас — картинки и тексты.

Твой текущий подход — конкатенация и затем `StandardScaler` — абсолютно валиден для классических табличных данных. Но для эмбеддингов он может быть неоптимален. `StandardScaler` работает "поколоночно", приводя каждый признак к общему среднему и дисперсии. 


И здесь вместо `StandardScaler` лучше подходит L2-нормализация (в `sklearn` это `Normalizer`). Этот метод масштабирует каждый вектор (каждую строку целиком) к единичной длине. Он убирает влияние "магнитуды" вектора (которая часто является шумом) и оставляет только его **направление** в многомерном пространстве — а именно оно и несет основную семантическую информацию.
    
Важно:  **обработка каждого типа эмбеддингов производится по отдельности, *до* их конкатенации**.
    
Таким образом, более продвинутый пайплайн мог бы выглядеть так:
1.  Получаем матрицу эмбеддингов для изображений `X_img`.
2.  Получаем матрицу эмбеддингов для текстов `X_text`.
3.  Применяем `Normalizer()` к `X_img` и `X_text` **по отдельности**.
4.  Конкатенируем уже нормализованные матрицы.

Такой подход гарантирует, что и визуальные, и текстовые признаки вносят в модель сопоставимый вклад, не искажая при этом структуру каждого из исходных пространств. Это знание будет очень полезно тебе в будущих проектах.
    

</div>

Отлично, исходная точка определена. Теперь перейдём к построению полносвязной нейронной сети.

In [38]:
X_train = torch.FloatTensor(np.array(X_train))
X_test = torch.FloatTensor(np.array(X_test))
y_train = torch.FloatTensor(np.array(y_train))
y_test = torch.FloatTensor(np.array(y_test))

In [39]:
X_train.shape

torch.Size([2233, 1466])

In [40]:
class Net(nn.Module):
    def __init__(self, n_in_neurons, n_hidden_neurons_1, n_hidden_neurons_2, n_hidden_neurons_3, n_hidden_neurons_4, n_out_neurons):
            super(Net, self).__init__()
            
            self.fc1 = nn.Linear(n_in_neurons, n_hidden_neurons_1)
            self.act1 = nn.ReLU()
            #self.do1 = nn.Dropout(p=0.7)
            self.fc2 = nn.Linear(n_hidden_neurons_1, n_hidden_neurons_2)
            self.act2 = nn.ReLU()
            self.fc3 = nn.Linear(n_hidden_neurons_2, n_hidden_neurons_3)
            self.act3 = nn.ReLU()
            self.fc4 = nn.Linear(n_hidden_neurons_3, n_hidden_neurons_4)
            self.act4 = nn.ReLU()
            self.fc5 = nn.Linear(n_hidden_neurons_4, n_out_neurons)
            self.act5 = nn.Sigmoid()
            
            nn.init.kaiming_uniform_(self.fc1.weight, mode='fan_in', nonlinearity='relu')
            nn.init.kaiming_uniform_(self.fc2.weight, mode='fan_in', nonlinearity='relu')
            nn.init.kaiming_uniform_(self.fc3.weight, mode='fan_in', nonlinearity='relu')
            nn.init.kaiming_uniform_(self.fc4.weight, mode='fan_in', nonlinearity='relu')
            nn.init.kaiming_uniform_(self.fc5.weight, mode='fan_in', nonlinearity='sigmoid')
            
    def forward(self, x):
        x = self.fc1(x)
        x = self.act1(x)
        #x = self.do1(x)
        x = self.fc2(x)
        x = self.act2(x)
        x = self.fc3(x)
        x = self.act3(x)
        x = self.fc4(x)
        x = self.act4(x)
        x = self.fc5(x)
        x = self.act5(x)
        
        return x

 <img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">


**Deep Learning: Рекомендации по архитектуре и обучению**

Ильгар, помимо исправления логики с признаками, давай "прокачаем" саму нейросеть. Сейчас она построена немного интуитивно, но в индустрии есть стандарты, которые помогают сетям учиться быстрее и стабильнее.

1. Магия степени двойки ()

У тебя используются размеры слоев: `1200`, `600`, `200`, `10`.
Компьютерное железо (GPU и память) оптимизировано для работы с числами, кратными степени двойки.
**Совет:** Используй стандартные размеры: `1024`, `512`, `256`, `128`, `64`.
Это касается и **Batch Size**: вместо `150` лучше взять `128` или `256`. Это мелочь, но она показывает твою "инженерную культуру".

2. Функция потерь: Почему MSE — это сложно?

Ты используешь `MSELoss` (среднеквадратичная ошибка) для задачи, где выход — это вероятность совпадения (от 0 до 1).

* **Проблема:** В паре с активацией `Sigmoid` на выходе, `MSE` часто приводит к "затуханию градиентов". Когда сеть ошибается сильно (выдает 0, когда надо 1), градиент может быть очень маленьким, и сеть перестает учиться.

* **Решение:** Используй **`nn.BCEWithLogitsLoss()`**.
* Это комбинация `Sigmoid` + `Binary Cross Entropy` в одной формуле.
* Она численно более стабильна и дает более сильные градиенты.
* *Важно:* Если используешь `BCEWithLogitsLoss`, убери `Sigmoid` из последнего слоя самой модели (функция сделает это внутри себя).
А вот для инференса (тестирования) уже придется добавить сигмоиду "руками".


3. Регуляризация (Борьба с переобучением)

В коде я вижу закомментированный `Dropout(p=0.7)`.
Нейросети — мастера "зубрежки". Без регуляризации они просто запомнят обучающую выборку.
**Совет:**

1. Верни **Dropout** (обычно ставят `p=0.2` - `0.5`), лучше между все всеми полносвязными слоями.
2. Используй батч-нормализацию




4. Мониторинг обучения (Слепые зоны)

Ты выводишь RMSE, но не строишь графики.
**Совет:** Обязательно сохраняй историю лоссов для **Train** и **Validation** на каждой эпохе и строй график двух линий.

* Если Train падает, а Val растет — **Переобучение** (нужен Dropout/ранняя остановка).
* Если оба высокие — **Недообучение** (нужна модель мощнее).
Без этого графика ты летишь "вслепую".

 5. Продвинутые техники (Schedulers & Early Stopping)

Обучать фиксированные 900 эпох — неэффективно.

* **Early Stopping:** Если лосс на валидации не падает в течение N эпох — останавливай обучение и сохраняй *лучшую* модель (а не последнюю).
* **LRScheduler:** Если обучение застряло (вышло на плато), полезно уменьшить шаг обучения (`learning rate`).
```python
scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer, mode='min', factor=0.1, patience=10)
# В цикле обучения:
# scheduler.step(val_loss)

```


Это позволит сети "скатиться" в более глубокий минимум функции потерь.
</div>


<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">


Для осознанной настройки нейронки важно понимать, как менять её архитектуру (усложнять и уменьшать регуляризацию? уменьшать шаг обучения? усиливать регуляризацию?). Для того чтобы принять эти  решения мало видеть только конечную цифру, важно понимать, чего модели не хватает чтобы метрика стала лучше.
    
    
Вот типичные  проблемы, которые можно выявить  анализируя графики и пути их решения:

<b>Недообучение (Underfitting)</b>
  Признаки: Высокие значения потерь и низкая точность как на обучающей, так и на валидационной выборке. Графики потерь не снижаются (или снижаются  недостаточно).
  
  Решения:
    
 - Увеличение сложности модели (добавление слоев, увеличение числа нейронов).
 - Использование более сложных моделей, таких как более глубокие нейронные сети.
 - Увеличение количества эпох обучения.

<b>Переобучение (Overfitting)</b>
  Признаки: Значительное снижение потерь и увеличение точности на обучающей выборке, но высокие потери и низкая точность на валидационной выборке. Графики начинают расходиться после некоторого количества эпох.
    
 Решения:
  - Регуляризация (L1, L2 регуляризация).
  - Dropout (добавление Dropout слоев).
  - Уменьшение сложности модели (уменьшение числа слоев или нейронов).
  - Использование аугментации данных для увеличения разнообразия данных.
  - Раннее завершение обучения (Early Stopping) на основе валидационной ошибки.

<b>Плохой выбор гиперпараметров</b>
 Признаки: Нестабильные графики потерь и точности (большие колебания).
 
 Решения:
  - Настройка скорости обучения (learning rate).
  - Оптимизация размера батча (batch size).
  - Попробовать другие оптимизаторы (например, Adam, SGD с различными параметрами).    

</div>

In [41]:
n_in_neurons = 2805
n_hidden_neurons_1 = 1200
n_hidden_neurons_2 = 600
n_hidden_neurons_3 = 200
n_hidden_neurons_4 = 10
n_out_neurons = 1



net = Net(n_in_neurons, n_hidden_neurons_1, n_hidden_neurons_2, n_hidden_neurons_3, n_hidden_neurons_4, n_out_neurons)

#net = nn.Sequential(
#    nn.Linear(n_in_neurons, n_hidden_neurons_1),
#    nn.ReLU(),
#    nn.Linear(n_hidden_neurons_1, n_hidden_neurons_2),
#    nn.ReLU(),
#    nn.Linear(n_hidden_neurons_2, n_out_neurons)
#)

optimizer2 = torch.optim.SGD(net.parameters(), lr=1e-3, momentum=0.8)

optimizers = [optimizer2]
loss = nn.MSELoss()

In [49]:
class Net(nn.Module):
    def __init__(self, input_dim):
        super().__init__()
        self.fc1 = nn.Linear(input_dim, 1200)  # Размерность input_dim должна совпадать с входом
        self.act1 = nn.ReLU()
        self.fc2 = nn.Linear(1200, 1)

    def forward(self, x):
        x = self.fc1(x)
        x = self.act1(x)
        x = self.fc2(x)
        return x


In [50]:
input_dim = 1466  # размерность признаков
net = Net(input_dim)

In [51]:
from torch.utils.data import DataLoader, TensorDataset

X_train_t = torch.tensor(X_train, dtype=torch.float32)
X_test_t = torch.tensor(X_test, dtype=torch.float32)

if isinstance(y_train, torch.Tensor):
    y_train_t = y_train
else:
    y_train_t = torch.tensor(y_train.to_numpy(), dtype=torch.float32)

if isinstance(y_test, torch.Tensor):
    y_test_t = y_test
else:
    y_test_t = torch.tensor(y_test.to_numpy(), dtype=torch.float32)


train_dataset = TensorDataset(X_train_t, y_train_t)
train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)


# Тренировка модели
num_epochs = 900
num_batches = ceil(len(X_train_t) / batch_size)

for optimizer in optimizers:
    optimizer_rmse = []
    print(f'Оптимизатор {optimizer}:')

    for epoch in range(num_epochs):
        net.train()  # Переключаем в режим обучения
        optimizer.zero_grad()  # Обнуляем градиенты на каждом шаге

        # Перемешиваем данные для каждого батча
        for batch_idx, (X_batch, y_batch) in enumerate(train_loader):
            # Преобразуем входы и цели в нужный тип (float32) для модели
            X_batch = X_batch.float()
            y_batch = y_batch.float()

            # Forward pass
            preds = net.forward(X_batch).flatten()

            # Расчет потерь
            loss_value = loss(preds, y_batch)

            # Обратное распространение ошибки
            loss_value.backward()

            # Шаг оптимизации
            optimizer.step()

        # Оценка модели раз в 10 эпох или в последней эпохе
        if epoch % 10 == 0 or epoch == num_epochs - 1:
            with torch.no_grad():
                net.eval()  # Переключаем в режим оценки
                test_preds = net.forward(X_test_t).flatten()  # Предсказания на тесте
                ans = round(float(torch.sqrt(loss(test_preds, y_test_t))), 2)  # RMSE
                print(f'Метрика RMSE на {epoch} итерации =', ans)
                optimizer_rmse.append(ans)

    print()
    print(f'Лучшая метрика RMSE = {min(optimizer_rmse)} обнаружена на {optimizer_rmse.index(min(optimizer_rmse))} эпохе')
    print(f'Средняя метрика RMSE = {round(np.mean(optimizer_rmse), 2)}')
    print('---------------------------------------------------------------------------------------------', '\n')


/var/folders/4s/p21vxwg5679gpxq0_nchdg7m0000gn/T/ipykernel_63035/2693908127.py:3: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.detach().clone() or sourceTensor.detach().clone().requires_grad_(True), rather than torch.tensor(sourceTensor).
  X_train_t = torch.tensor(X_train, dtype=torch.float32)
/var/folders/4s/p21vxwg5679gpxq0_nchdg7m0000gn/T/ipykernel_63035/2693908127.py:4: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.detach().clone() or sourceTensor.detach().clone().requires_grad_(True), rather than torch.tensor(sourceTensor).
  X_test_t = torch.tensor(X_test, dtype=torch.float32)


Оптимизатор SGD (
Parameter Group 0
    dampening: 0
    differentiable: False
    foreach: None
    fused: None
    lr: 0.001
    maximize: False
    momentum: 0.8
    nesterov: False
    weight_decay: 0
):
Метрика RMSE на 0 итерации = 1.14
Метрика RMSE на 10 итерации = 1.14
Метрика RMSE на 20 итерации = 1.14
Метрика RMSE на 30 итерации = 1.14
Метрика RMSE на 40 итерации = 1.14
Метрика RMSE на 50 итерации = 1.14
Метрика RMSE на 60 итерации = 1.14
Метрика RMSE на 70 итерации = 1.14
Метрика RMSE на 80 итерации = 1.14
Метрика RMSE на 90 итерации = 1.14
Метрика RMSE на 100 итерации = 1.14
Метрика RMSE на 110 итерации = 1.14
Метрика RMSE на 120 итерации = 1.14
Метрика RMSE на 130 итерации = 1.14
Метрика RMSE на 140 итерации = 1.14
Метрика RMSE на 150 итерации = 1.14
Метрика RMSE на 160 итерации = 1.14
Метрика RMSE на 170 итерации = 1.14
Метрика RMSE на 180 итерации = 1.14
Метрика RMSE на 190 итерации = 1.14
Метрика RMSE на 200 итерации = 1.14
Метрика RMSE на 210 итерации = 1.14
Метрика RMS

## 7. Тестирование модели

Настало время протестировать модель. Для этого получите эмбеддинги для всех тестовых изображений из папки `test_images`, выберите случайные 10 запросов из файла `test_queries.csv` и для каждого запроса выведите наиболее релевантное изображение. Сравните визуально качество поиска.

Для начала сформируем массив эмбеддингов. Для этого воспользуемся ранее реализованной функцией vect_gen_train.

In [52]:
path="to_upload/test_images/"
test_images_vec = np.vstack(np.array(test_images.apply(vect_gen_train, axis=1)))

Реализуем функцию, возвращающую список из десяти случайно выбранных описаний.

In [53]:
def ten_texts():
    texts = []
    for i in range(0,10):
        texts.append(test_queries.iloc[rd.randint(0, 499)]['query_text'])
    return texts

Теперь напишем функцию, которая подготовит текстовое описание ко входу в модель. Для этого тестовое описание векторизуется и соединяется с каждым из подготовленых изображений. За одно проверим тут текстовый запрос на наличие "запретных" слов

In [54]:
def conv_text(text):
    corpus_one = lemma_clear(text)
    corpus = []
    #print(corpus_one)
    #print(is_ban_word(corpus_one))
    if not is_ban_word(corpus_one):
        return None
    else:
        #tf_idf_one = count_tf_idf.transform(corpus)
        for i in range(0,100):
            corpus.append(corpus_one)
        tf_idf_one = count_tf_idf.transform(corpus)
        return(np.hstack([np.vstack(test_images_vec), tf_idf_one.toarray()]))

Теперь можно перейти к этапу тестирования. Для каждого из 10 отобранных описаний будет выведено само описание и изображение, набравшее наибольшее количество «очков».

In [55]:
for item in ten_texts():
    print(item)
    try:
        mass = conv_text(item)
        res = net.forward(torch.FloatTensor(mass)).flatten().detach().numpy()
        print(res.argmax())
    except:
        print('Наличие запретного слова в запросе!')

Two boys spraying each other with water
Наличие запретного слова в запросе!
Two black dogs are fighting over a toy
43
The man is dressed in black and smiling while another man is drinking and an elderly woman is standing near a fence .
43
Two children play with a white dog .
Наличие запретного слова в запросе!
A little girl buried in the sand .
Наличие запретного слова в запросе!
Three people , one wearing a t-shirt and baseball cap , are seated on a bus .
43
A family watches as a little girl blows out the candles on her third birthday
Наличие запретного слова в запросе!
At the bottom of a set of stairs , there are two men and a tent .
43
A white dog with brown ears is running on the sidewalk .
43
A woman and two children pose for a picture under some rocks .
Наличие запретного слова в запросе!


 <img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">

Скорее всего проблема `предсказание одного изображения` возникает из-за использования TF-IDF векторизации. Сильно разный масштаб признаков + у TF-IDF признаки разряженные (большая часть  нули и очень мало отличаются от нуля) приводит к тому, что модель просто не замечает эти признакии и учится работать исключительно с картинками.
    
    
Выше я оставил советы, надеюсь они помогут.
    
    
Кстати, я замечал что с TF-IDF неплохо справляются бустинги. Можешь попробовать эти модели, как минимум для эксперимента. 
    
    
</div>


Поскольку модель во всех случаях выдаёт одно и то же изображение, проведение визуальной оценки качества предсказаний не представляется целесообразным. Решить данную проблему самостоятельно не удалось. В связи с этим по рекомендации преподавателя работа была направлена на ревью с целью получения обратной связи и рекомендаций по возможным способам устранения данной проблемы.

## 8. Выводы

- [x]  Jupyter Notebook открыт
- [ ]  Весь код выполняется без ошибок
- [ ]  Ячейки с кодом расположены в порядке исполнения
- [ ]  Исследовательский анализ данных выполнен
- [ ]  Проверены экспертные оценки и краудсорсинговые оценки
- [ ]  Из датасета исключены те объекты, которые выходят за рамки юридических ограничений
- [ ]  Изображения векторизованы
- [ ]  Текстовые запросы векторизованы
- [ ]  Данные корректно разбиты на тренировочную и тестовую выборки
- [ ]  Предложена метрика качества работы модели
- [ ]  Предложена модель схожести изображений и текстового запроса
- [ ]  Модель обучена
- [ ]  По итогам обучения модели сделаны выводы
- [ ]  Проведено тестирование работы модели
- [ ]  По итогам тестирования визуально сравнили качество поиска